conda-store-server/tests/user_journeys/helpers/helpers.py

```diff
@@ -0,0 +1,145 @@
+        data = token_response.json()
+        self.token = data["data"]["token"]
+
+
+def get_current_time() -> datetime:
+    """ Get the current time. """
+    return datetime.now(timezone.utc)
+
+
+def get_time_in_future(hours: int) -> datetime:
+    """ Get the time in the future."""
+    current_time = get_current_time()
+    future_time = current_time + timedelta(hours=hours)
+    return future_time
+
+
+def get_iso8601_time(hours: int) -> str:
+    """ Get the time in the future in ISO 8601 format. """
+    future_time = get_time_in_future(hours)
+    iso_format = future_time.isoformat()
+    return iso_format
+
+
+def create_namespace(api: APIHelper, namespace: str) -> requests.Response:
+    """ Create a namespace. """
+    return api.make_request(
+        f"api/v1/namespace/{namespace}", method="POST")
+
+
+def create_token(
+        api: APIHelper, namespace: str, role: str,
+        default_namespace: str = "default"
+        ) -> requests.Response:
+    """ Create a token with a specified role in a specified namespace. """
+    one_hour_in_future = get_time_in_future(1)
+    json_data = {
+        "primary_namespace": default_namespace,
+        "expiration": one_hour_in_future.isoformat(),
+        "role_bindings": {
+            f"{namespace}/*": [role]
+        }
+    }
+    return api.make_request(
+        "api/v1/token", method="POST", json_data=json_data)
+
+
+def create_environment(
+        api: APIHelper,  namespace: str, specification_path: str
+        ) -> requests.Response:
+    """
+        Create an environment.
+        The environment specification is read
+        from a conda environment.yaml file.
+    """
+    with open(specification_path, "r", encoding="utf-8") as file:
+        specification_content = file.read()
+
+    json_data = {
+        "namespace": namespace,
+        "specification": specification_content
+    }
+
+    return api.make_request(
+        "api/v1/specification", method="POST", json_data=json_data)
+
+
+def wait_for_successful_build(
+        api: APIHelper, build_id: str
+        ) -> requests.Response:
+    """ Wait for a build to complete."""
+    status = APIBuildStatus.QUEUED.value
+    while status != APIBuildStatus.COMPLETED.value:
+        response = api.make_request(
+            f"api/v1/build/{build_id}", method="GET")
+        status = response.json()["data"]["status"]
+        if status == APIBuildStatus.FAILED.value:
+            raise AssertionError("Build failed")
+    return response
+
+
+def delete_environment(
+        api: APIHelper, namespace: str, environment_name: str
+        ) -> requests.Response:
+    """ Delete an environment."""
+    return api.make_request(
+        f"api/v1/environment/{namespace}/{environment_name}",
+        method="DELETE")
+
+
+def delete_namespace(api: APIHelper, namespace: str) -> requests.Response:
+    """ Delete a namespace."""
+    return api.make_request(f"api/v1/namespace/{namespace}", method="DELETE")
```

Comment:
I really prefer smaller classes with just the class attributes and methods that operate on the attributes. I had originally written it as one huge class, but then refactored so the class was much smaller and the other functions could just use the `make_request` method. However, I am not going to argue too hard on that point, so I went ahead and moved it back to what I had before.

Reply:
okay, I have been thinking about this more, and I want to expand on my thoughts above about why I chose the approach of the smaller class. I understand convention should be followed within a code base, but the huge classes is something that does bother me about this code base, and makes testing much harder imo.

In my original implementation, the `API` class did just a few things. It stored a base url and token which were needed to connect to the server, and had a `make_request` method which used the url and token to make api calls. That is all good. Classes store data and methods that use that data, and in this case every call would need both the base url and token.

However, there is nothing specific to a particular implementation for functions such as create_environment or delete_environment. As long are passed an `API` that knows how to connect, they don't care about the details. This isn't a huge deal in test code like this, but in the application code this architecture makes testing and mocking far easier. We could even go farther and have `Requester` protocol class with nothing but a `make_request` method that the `API` class implements which would make mocking requests and testing far easier and help the maintainability of the code.

Like I said earlier, I recognize that following conventions within code bases is important for readability, but I think we should be moving towards smaller classes with less responsibility so that I don't have to instatiate a whole API class to test a single method.